Full name: Vũ Minh Phát

Student ID: 21127739

# HW4: Parallel Radix Sort

**To compile your file, you can use this command:** \
`nvcc tên-file.cu -o tên-file-chạy` \
***You can use Vietnamese to anwser the questions***

**Đề bài**: Song song hóa Radix Sort, trong đó bước scan viết trong một hàm kernel duy nhất (xem file slide "`CSC14120-Lecture09-RadixSort.pdf`").

Biên dịch file code `HW4.cu` bằng trình biên dịch `nvcc`:

In [1]:
!nvcc HW4.cu -o HW4

Tiếp theo, ta sẽ chạy chương trình với các kích thước block khác nhau để tìm ra kích thước block tối ưu nhất. Với mỗi kích thước block, ta sẽ chạy file `HW4` theo cả 2 cách:
- Cách 1: Chạy file `HW4` một cách thông thường. Sử dụng câu lệnh: `./HW4 <block_size>`
- Cách 2: Dùng thêm `nvprof` để xem chi tiết thực thi của chương trình. Sử dụng câu lệnh: `nvprof ./HW4 <block_size>`

Ta sẽ dùng kết quả của **cách 1** để tính ra thời gian thực thi của chương trình ứng với mỗi kích thước block. Do đó, ta sẽ chạy code cell này nhiều lần để thu được kết quả chính xác nhất. Còn đối với **cách 2**, ta sẽ quan tâm đến thời gian thực thi các công việc trong phần "GPU activities". Thông tin này sẽ giúp ta hiểu rõ hơn về tác động của kích thước block đến hiệu suất của chương trình.

Bây giờ, ta sẽ bắt đầu chạy file `HW4` với các kích thước block khác nhau:

- Với kích thước block là `256`

In [4]:
!./HW4 256

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 byte
SMEM per SM: 65536 byte
SMEM per block: 49152 byte
****************************

Input size: 16777217

Radix Sort by host
Time: 9760.913 ms

Radix Sort by device
Block size: 256, Grid size: 32769
Time: 915.423 ms
CORRECT :)


In [5]:
!nvprof ./HW4 256

==564== NVPROF is profiling process 564, command: ./HW4 256
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 byte
SMEM per SM: 65536 byte
SMEM per block: 49152 byte
****************************

Input size: 16777217

Radix Sort by host
Time: 9411.372 ms

Radix Sort by device
Block size: 256, Grid size: 32769
Time: 918.616 ms
CORRECT :)
==564== Profiling application: ./HW4 256
==564== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   87.14%  791.14ms        32  24.723ms  20.240ms  47.675ms  BrentKungExclusiveScanKernel(int*, int*, int, int*)
                    5.41%  49.119ms         1  49.119ms  49.119ms  49.119ms  [CUDA memcpy DtoH]
                    3.95%  35.852ms        32  1.1204ms  1.0462ms  1.1295ms  computeRankAndWriteToDstKernel(unsigned int*, int*, int*, unsigned int*, int)
                    1.92%  17.

- Với kích thước block là `512`

In [7]:
!./HW4 512

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 byte
SMEM per SM: 65536 byte
SMEM per block: 49152 byte
****************************

Input size: 16777217

Radix Sort by host
Time: 9446.203 ms

Radix Sort by device
Block size: 512, Grid size: 16385
Time: 573.924 ms
CORRECT :)


In [8]:
!nvprof ./HW4 512

==783== NVPROF is profiling process 783, command: ./HW4 512
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 byte
SMEM per SM: 65536 byte
SMEM per block: 49152 byte
****************************

Input size: 16777217

Radix Sort by host
Time: 9470.033 ms

Radix Sort by device
Block size: 512, Grid size: 16385
Time: 570.200 ms
CORRECT :)
==783== Profiling application: ./HW4 512
==783== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   79.47%  445.67ms        32  13.927ms  10.120ms  23.764ms  BrentKungExclusiveScanKernel(int*, int*, int, int*)
                    8.19%  45.929ms         1  45.929ms  45.929ms  45.929ms  [CUDA memcpy DtoH]
                    6.37%  35.735ms        32  1.1167ms  1.0344ms  1.1234ms  computeRankAndWriteToDstKernel(unsigned int*, int*, int*, unsigned int*, int)
                    3.19%  17.

- Với kích thước block là `1024`

In [10]:
!./HW4 1024

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 byte
SMEM per SM: 65536 byte
SMEM per block: 49152 byte
****************************

Input size: 16777217

Radix Sort by host
Time: 9453.724 ms

Radix Sort by device
Block size: 1024, Grid size: 8193
Time: 428.116 ms
CORRECT :)


In [12]:
!nvprof ./HW4 1024

==1205== NVPROF is profiling process 1205, command: ./HW4 1024
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 byte
SMEM per SM: 65536 byte
SMEM per block: 49152 byte
****************************

Input size: 16777217

Radix Sort by host
Time: 9529.373 ms

Radix Sort by device
Block size: 1024, Grid size: 8193
Time: 399.449 ms
CORRECT :)
==1205== Profiling application: ./HW4 1024
==1205== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   69.62%  271.69ms        32  8.4903ms  5.5262ms  11.982ms  BrentKungExclusiveScanKernel(int*, int*, int, int*)
                   12.26%  47.836ms         1  47.836ms  47.836ms  47.836ms  [CUDA memcpy DtoH]
                    9.31%  36.318ms        32  1.1349ms  1.0386ms  1.1432ms  computeRankAndWriteToDstKernel(unsigned int*, int*, int*, unsigned int*, int)
                    4.80

**Câu hỏi**: Giải thích tại sao khi thay đổi kích thước block thì kết quả lại thay đổi như vậy?

---

**`1.` Quan sát sự thay đổi của kết quả khi thay đổi kích thước block và đưa ra một số nhận xét:**

Đầu tiên, ta sẽ tạo ra một bảng thống kê thời gian chạy của chương trình ứng với các kích thước block khác nhau. Với mỗi kích thước block, ta chạy chương trình `5 lần`, sau đó tính ra thời gian chạy trung bình. Trước khi bắt đầu ghi nhận thời gian, ta sẽ chạy chương trình một lần để "làm nóng" GPU.

| Kích thước block (Block size) | Số lượng block (Grid size) | Chạy lần 1 (ms) | Chạy lần 2 (ms) | Chạy lần 3 (ms) | Chạy lần 4 (ms) | Chạy lần 5 (ms) | Thời gian chạy trung bình (ms) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 256 | 32769 | 940.082 | 945.507 | 907.579 | 917.287 | 940.38 | 930.167 |
| 512 | 16385 | 596.125 | 580.446 | 590.943 | 598.832 | 568.071 | 586.883 |
| 1024 | 8193 | 424.264 | 439.849 | 397.475 | 420.168 | 430.14 | **422.379** |

- **Nhận xét**:

    - Từ bảng thống kê thời gian chạy của chương trình ứng với các kích thước block khác nhau, ta phát hiện có một xu hướng rõ rệt: *thời gian chạy của chương trình giảm dần khi kích thước block tăng lên*. Điều này được thể hiện rõ qua bảng thống kê, với thời gian chạy trung bình giảm từ 930.167ms (block size 256) xuống còn 422.379ms (block size 1024).

    - Ta sẽ tìm hiểu nguyên nhân dẫn đến hiện tượng này bằng cách phân tích chi tiết hơn về các hoạt động của GPU trong khi thực thi chương trình.

Quan sát thời gian thực thi các công việc trong phần "GPU activities" của `nvprof` ứng với mỗi kích thước block, ta phát hiện rằng thời gian thực thi của các công việc:

- Sao chép dữ liệu từ host sang device (HtoD)
- Sao chép dữ liệu từ device sang host (DtoH)
- Bước rút trích bit
- Bước tính rank và ghi kết quả vào mảng dst

không thay đổi nhiều khi ta thay đổi kích thước block. Nghĩa là, thời gian thực thi của các công việc này không phụ thuộc quá nhiều vào kích thước block.

| Công việc | Thời gian chạy (ms) |
| --- | --- |
| [CUDA memcpy HtoD] | 15 |
| [CUDA memcpy DtoH] | 46 |
| extractBitsKernel | 18 |
| computeRankAndWriteToDstKernel | 36 |

Tuy nhiên, thời gian thực thi của bước "exclusive scan" để tính ra mảng `nOnesBefore` lại có xu hướng giảm xuống khi ta tăng kích thước block. Và chính sự khác biệt trong thời gian thực thi của hàm kernel này đã ảnh hưởng trực tiếp đến tổng thời gian chạy của chương trình ứng với các kích thước block khác nhau.

| Kích thước block (Block size) | Thời gian chạy của bước "exclusive scan" (ms) |
| --- | --- |
| 256 | 809.15 |
| 512 | 451.90 |
| 1024 | **297.94** |

- **Nhận xét**:

    - Sau khi tiến hành phân tích chi tiết hơn về thời gian thực thi của các công việc có sử dụng GPU, ta nhận thấy rằng nguyên nhân chính dẫn đến sự thay đổi thời gian chạy của chương trình khi thay đổi kích thước block nằm ở bước "exclusive scan" để tính mảng `nOnesBefore`. Trong khi thời gian thực thi của các công việc khác như sao chép dữ liệu và các bước kernel khác ít thay đổi, thì thời gian thực thi của bước "exclusive scan" giảm mạnh khi kích thước block tăng lên (từ 809.15ms xuống 297.94ms).

    - Tiếp theo, ta sẽ phân tích chi tiết về mỗi lần lặp của thuật toán Radix Sort, cũng như cách hoạt động của bước "exclusive scan" để hiểu rõ hơn về nguyên nhân dẫn đến sự thay đổi này.

**`2.` Phân tích chi tiết về mỗi lần lặp của thuật toán Radix Sort và bước "exclusive scan":**

Trong mỗi lần lặp (iteration) của thuật toán Radix Sort, ta cần phải thực hiện 3 bước chính:

- Bước 1: Rút trích bit thứ `bitIdx` từ mỗi phần tử của mảng `src` và ghi vào mảng `bits`.

- Bước 2: Dựa vào mảng `bits`, ta tính ra mảng `nOnesBefore` (số lượng bit 1 trước mỗi phần tử của mảng `src`).

- Bước 3: Dựa vào mảng `bits` và `nOnesBefore`, ta sắp xếp lại vị trí của các phần tử trong mảng `src` và ghi kết quả vào mảng `dst`. Vị trí mới của mỗi phần tử trong mảng `src` sẽ được tính dựa vào số lượng bit 0 và bit 1 trước nó (xem công thức tính `rank` trong slide).

Cả 3 bước này đều có thể thực hiện song song trên GPU thông qua việc sử dụng các hàm kernel. Trong đó, với "**bước 1**" và "**bước 3**", các thread block có thể hoạt động độc lập với nhau, tức là các block không cần chờ đợi tín hiệu từ các block khác. Do đó, sự thay đổi về số lượng block cần thiết trong grid không ảnh hưởng nhiều đến hiệu suất của chương trình.

Ngược lại, ở "**bước 2**", ta cần phải thực hiện một **thuật toán "exclusive scan"** song song trên toàn bộ grid, tức là kết quả của mỗi block sẽ có sự phụ thuộc vào kết quả của các block "trước đó". Thuật toán "exclusive scan" này có thể được thực hiện thông qua 4 bước chính:

- Bước 2.1: Tính ra block index $bi$ của mỗi block theo thứ tự được đưa vào SM của chúng.

- Bước 2.2: Thực hiện scan cục bộ trên mỗi block để tính ra số lượng bit 1 trước mỗi phần tử trong block đó. Đồng thời, ta cũng cần phải lưu lại tổng số lượng bit 1 của mỗi block vào một mảng riêng để sử dụng cho bước tiếp theo.

- Bước 2.3: Block $bi$ sẽ phải chờ đến khi block $bi-1$ báo tín hiệu rằng nó đã hoàn thành việc tính tổng toàn cục của $bi$ block trước đó (từ block 0 đến block $bi-1$). Sau đó, block $bi$ sẽ cộng "tổng toàn cục của các block trước đó" vào "tổng cục bộ của mình" và truyền tín hiệu cho block $bi+1$ biết rằng nó có thể bắt đầu công việc của mình. 

- Bước 2.4: Thêm tổng toàn cục của $bi$ block trước đó vào kết quả của mỗi phần tử trong block $bi$.

Trong 4 bước chính được liệt kê ở trên, thì chỉ có "**bước 2.2**" và "**bước 2.4**" là được thực hiện song song một cách "tự nhiên", tức là hoạt động của mỗi thread block không phụ thuộc vào (hoặc không cần chờ đợi) các block khác. Còn "**bước 2.1**" và "**bước 2.3**" lại đòi hỏi sự phối hợp giữa các block, tức là các thread block phải làm việc một cách tuần tự, block $bi$ phải chờ đến khi block $bi-1$ bật tín hiệu "hoàn thành" thì mới được thực hiện bước tiếp theo.

**`3.` Giải thích về ảnh hưởng của kích thước block đến hiệu suất của chương trình:**

Lý do cho sự thay đổi về thời gian chạy của chương trình khi thay đổi kích thước block nằm ở cách thuật toán "exclusive scan" được thực hiện trên GPU. Như đã phân tích, thuật toán này bao gồm các bước cần sự phối hợp giữa các block, đặc biệt là "**bước 2.3**", nơi block $bi$ phải chờ block $bi-1$. **Khi kích thước block tăng lên, thì số lượng block cần thiết để xử lý cùng một lượng dữ liệu giảm xuống**. Điều này dẫn đến:

`3.1.` **Giảm số lượng bước đồng bộ hóa giữa các block**:

- Khi kích thước block nhỏ, ta sẽ cần nhiều block hơn để xử lý toàn bộ dữ liệu, dẫn đến nhiều lần đồng bộ hóa giữa các block. Điều này gây ra overhead lớn hơn. Khi tăng kích thước block, số lượng block cần thiết để xử lý cùng một khối lượng dữ liệu sẽ giảm, từ đó giảm số lần đồng bộ hóa giữa các block. Đây chính là yếu tố then chốt giúp giảm thời gian thực thi. Việc đồng bộ hóa giữa các block là một hoạt động tốn kém, và việc giảm thiểu nó sẽ cải thiện hiệu suất đáng kể. Điều này giải thích tại sao thời gian thực thi của bước "exclusive scan" giảm mạnh khi kích thước block tăng lên. 

`3.2.` **Tăng hiệu quả sử dụng bộ nhớ chia sẻ (shared memory)**:

- Mỗi block sử dụng shared memory để lưu trữ dữ liệu trung gian. Với kích thước block lớn hơn, bộ nhớ chia sẻ có thể được sử dụng hiệu quả hơn vì nhiều dữ liệu có thể được xử lý cục bộ mà không cần truy cập nhiều vào bộ nhớ toàn cục (global memory). Điều này giảm tải băng thông bộ nhớ toàn cục, giúp cải thiện tốc độ truy cập dữ liệu.

`3.3.` **Tăng hiệu suất tính toán cục bộ**:

- Kích thước block lớn hơn cho phép mỗi block xử lý nhiều dữ liệu hơn ở cục bộ. Điều này có thể tận dụng tốt hơn bộ nhớ shared memory của SM (Streaming Multiprocessor) và giảm thiểu số lần truy cập bộ nhớ global memory, vốn chậm hơn. Bước 2.2 và 2.4 (tính toán cục bộ trong mỗi block) sẽ hiệu quả hơn với block size lớn hơn.

**Kết luận**:

- Sự giảm thời gian chạy khi tăng kích thước block là do GPU có thể tận dụng tốt hơn khả năng xử lý song song và bộ nhớ chia sẻ, tối ưu hóa việc sử dụng băng thông bộ nhớ.

- Đặc biệt, việc tăng kích thước block giúp giảm số lượng block cần thiết, từ đó giảm số lần đồng bộ hóa giữa các block trong bước "exclusive scan", dẫn đến giảm đáng kể thời gian thực thi của bước này và do đó giảm tổng thời gian chạy của chương trình.

- Điều này cho thấy việc lựa chọn kích thước block phù hợp là rất quan trọng để tối ưu hiệu năng của các thuật toán song song trên GPU, đặc biệt là các thuật toán có bước cần sự phối hợp giữa các block như thuật toán "exclusive scan" mà chúng ta đã cài đặt.


> References:  
> [1] Book "Programming Massively Parallel Processors: A Hands-on Approach", chapter "11 - Prefix sum (scan)", by David B. Kirk and Wen-mei W. Hwu.  
> [2] Book "Programming Massively Parallel Processors: A Hands-on Approach", chapter "13 - Sorting", by David B. Kirk and Wen-mei W. Hwu.  
> [3] Slide bài giảng "2023-CSC14120-Lecture08-Scan.pdf" được cung cấp trong môn học.  
> [4] Slide bài giảng "2023-CSC14120-Lecture09-RadixSort.pdf" được cung cấp trong môn học.  
> [5] File code "08-Scan.cu" trong phần Demo code trên drive được cung cấp trong môn học.  
> [6] "Chapter 39. Parallel Prefix Sum (Scan) with CUDA" in "GPU Gems 3" by Mark Harris, Shubhabrata Sengupta, John D. Owens - [developer.nvidia.com](https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda).  